File: python/admitiq/stores/hosted_store.py
"""
admitiq.stores.hosted_store — optional hosted revocation store.

Calls the AdmitiQ hosted API for multi-scanner single-use without Redis.
The core library never requires this service.
"""
from __future__ import annotations

import json
import urllib.error
import urllib.request
from typing import Any, Optional


class HostedRevocationStore:
    """
    Usage:
        store = HostedRevocationStore(api_key=os.environ["ADMITIQ_API_KEY"])
        payload = verify(token, secret=SECRET, is_revoked=store.is_revoked)
        store.mark_used(payload["jti"])
    """

    def __init__(
        self,
        api_key: str,
        base_url: str = "https://api.admitiq.logiclitz.org",
        timeout: float = 10.0,
    ):
        if not api_key:
            raise ValueError("HostedRevocationStore requires api_key")
        self._api_key = api_key
        self._base_url = base_url.rstrip("/")
        self._timeout = timeout

    def _post(self, path: str, body: dict) -> dict:
        req = urllib.request.Request(
            f"{self._base_url}{path}",
            data=json.dumps(body).encode("utf-8"),
            headers={
                "Authorization": f"Bearer {self._api_key}",
                "Content-Type": "application/json",
            },
            method="POST",
        )
        try:
            with urllib.request.urlopen(req, timeout=self._timeout) as resp:
                return json.loads(resp.read().decode("utf-8"))
        except urllib.error.HTTPError as e:
            raw = e.read().decode("utf-8", errors="replace")
            try:
                data = json.loads(raw)
            except Exception:
                data = {"error": raw}
            raise RuntimeError(data.get("error") or f"hosted_http_{e.code}") from e

    def mark_used(self, jti: str) -> bool:
        data = self._post("/v1/revoke", {"jti": jti})
        return bool(data.get("first"))

    def is_revoked(self, jti: str) -> bool:
        data = self._post("/v1/check", {"jti": jti})
        return bool(data.get("revoked"))

File: js/src/stores/hostedStore.js
/**
 * admitiq/stores/hostedStore — optional hosted revocation store.
 *
 * Calls the AdmitiQ hosted API for multi-scanner single-use without Redis.
 * The core library never requires this.
 *
 *   const store = new HostedRevocationStore({ apiKey: process.env.ADMITIQ_API_KEY });
 *   await verify(token, SECRET, store.isRevoked.bind(store));
 *   await store.markUsed(payload.jti);
 */
class HostedRevocationStore {
  /**
   * @param {object} options
   * @param {string} options.apiKey
   * @param {string} [options.baseUrl="https://api.admitiq.logiclitz.org"]
   * @param {typeof fetch} [options.fetchImpl]
   */
  constructor({
    apiKey,
    baseUrl = "https://api.admitiq.logiclitz.org",
    fetchImpl = globalThis.fetch,
  } = {}) {
    if (!apiKey) throw new Error("HostedRevocationStore requires apiKey");
    if (!fetchImpl) throw new Error("HostedRevocationStore requires fetch (Node 18+)");
    this.apiKey = apiKey;
    this.baseUrl = baseUrl.replace(/\/$/, "");
    this.fetchImpl = fetchImpl;
  }

  async _post(path, body) {
    const res = await this.fetchImpl(`${this.baseUrl}${path}`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${this.apiKey}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(body),
    });
    const data = await res.json().catch(() => ({}));
    if (!res.ok) {
      const err = new Error(data.error || `hosted_http_${res.status}`);
      err.status = res.status;
      err.data = data;
      throw err;
    }
    return data;
  }

  async markUsed(jti) {
    const data = await this._post("/v1/tokens/consume", { jti });
    return Boolean(data.first);
  }

  async isRevoked(jti) {
    const data = await this._post("/v1/tokens/check", { jti });
    return Boolean(data.revoked);
  }

  async unrevoke(jti) {
    const data = await this._post("/v1/tokens/unrevoke", { jti });
    return Boolean(data.unrevoked);
  }
}

module.exports = { HostedRevocationStore };
